### PR TITLE
Create CModAdd

### PR DESCRIPTION
**Pull request by NoureldinYosri**
Controlled modular addition as it appears in https://arxiv.org/pdf/2306.08585

### CODE REVIEW

File: qualtran/bloqs/arithmetic/addition.py [hunk 1]
```diff
@@ -452,7 +452,7 @@ def build_composite_bloq(
         if self.signed:
             binary_rep = QInt(self.bitsize).to_bits(self.k)
         else:
-            binary_rep = QUInt(self.bitsize).to_bits(self.k)
+            binary_rep = QUInt(self.bitsize).to_bits(self.k % (2**self.bitsize))
```
Comment (mpharrigan) [inline qualtran/bloqs/arithmetic/addition.py:455]: what's going on here? should we raise an error if `self.k` is too big in the class constructor? What about the line two lines up; should that be modified as well? Are there other locations where we currently assume `self.k` fits into `self.bitsize`?

File: qualtran/bloqs/mod_arithmetic/mod_addition_test.py [hunk 2]
```diff
@@ -90,3 +97,51 @@ def test_classical_action_mod_add(prime, bitsize):
...
             assert b.call_classically(x=x, y=y) == cb.call_classically(x=x, y=y)
+
+
+@pytest.mark.parametrize('control', range(2))
+@pytest.mark.parametrize(
+    ['prime', 'bitsize'],
+    [(p, bitsize) for p in [11, 13, 31] for bitsize in range(1 + p.bit_length(), 8)],
+)
+@pytest.mark.parametrize('dtype', [QUInt, QMontgomeryUInt])
+def test_classical_action_cmodadd(control, prime, dtype, bitsize):
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition_test.py:108]: nice! can you mark this exhaustive test as `pytest.mark.slow` and craft a `_quick` version: pick specific prime,bitsize pairs; take a couple (seeded) random values for x and y in valid_range

File: qualtran/bloqs/mod_arithmetic/mod_addition_test.py [hunk 3]
```diff
@@ -90,3 +97,51 @@ def test_classical_action_mod_add(prime, bitsize):
...
+    b = CModAdd(dtype(bitsize), mod=prime, cv=control)
+    cb = b.decompose_bloq()
+    valid_range = range(prime)
+    for c in range(2):
+        for x in valid_range:
+            for y in valid_range:
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition_test.py:114]: optional nit: could use `itertools.product` to reduce levels of indentation

File: qualtran/bloqs/mod_arithmetic/mod_addition.py [hunk 4]
```diff
@@ -339,3 +342,109 @@ def wire_symbol(
         if reg.name == 'y':
             return TextBox(f'y += x*{self.k}')
         raise ValueError(f"Unknown register {reg}")
+
+
+@frozen
+class CModAdd(Bloq):
+    r"""Controlled Modular Addition.
+
+    Implements $\ket{c}\ket{x}\ket{y} \xrightarrow[]{} \ket{c}\ket{x}\ket{(cx+y)\%p}$
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition.py:351]: can you just use `\rightarrow`? lots of empty brackets happening in the current arrow

Does `\ket` show up in the notebooks?

File: qualtran/bloqs/mod_arithmetic/mod_addition.py [hunk 5]
```diff
@@ -339,3 +342,109 @@ def wire_symbol(
...
+    using $5n+1$ Toffoli gates.
+
+    Note: The true toffoli count is $5n$ however we have an extra toffoli
+    due to how our OutOfPlaceAdder works.
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition.py:355]: I'm not wild about the phrase "true toffoli count". Can you re-word this.

Note: the reference reports 5n toffolis. Our construction has an extra toffoli gate due to the current implementaiton of `OutOfPlaceAdder`

or something similar

File: qualtran/bloqs/mod_arithmetic/mod_addition_test.py [hunk 6]
```diff
@@ -90,3 +97,51 @@ def test_classical_action_mod_add(prime, bitsize):
     for x in valid_range:
         for y in valid_range:
             assert b.call_classically(x=x, y=y) == cb.call_classically(x=x, y=y)
+
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition_test.py:100]: can you patch in this test:

```patch
diff --git a/qualtran/bloqs/mod_arithmetic/mod_addition_test.py b/qualtran/bloqs/mod_arit>
index ab983743..4384c4c0 100644
--- a/qualtran/bloqs/mod_arithmetic/mod_addition_test.py
+++ b/qualtran/bloqs/mod_arithmetic/mod_addition_test.py
@@ -20,7 +20,7 @@ from qualtran.bloqs.arithmetic import Add
 from qualtran.bloqs.mod_arithmetic import CModAdd, CModAddK, CtrlScaleModAdd, ModAdd, Mo>
 from qualtran.bloqs.mod_arithmetic.mod_addition import _cmodadd_example
 from qualtran.cirq_interop.t_complexity_protocol import TComplexity
-from qualtran.resource_counting import GateCounts, QECGatesCost, query_costs
+from qualtran.resource_counting import GateCounts, get_cost_value, QECGatesCost, query_c>
 from qualtran.resource_counting.generalizers import ignore_alloc_free, ignore_split_join
 from qualtran.testing import (
     assert_equivalent_bloq_counts,
@@ -99,6 +99,18 @@ def test_classical_action_mod_add(prime, bitsize):
             assert b.call_classically(x=x, y=y) == cb.call_classically(x=x, y=y)
 
 
+def test_cmod_add_complexity_vs_ref():
+    n, k = sympy.symbols('n k', integer=True, positive=True)
+    bloq = CModAdd(QUInt(n), mod=k)
+    counts = get_cost_value(bloq, QECGatesCost()).total_t_and_ccz_count()
+    assert counts['n_t'] == 0, 'all toffoli'
+
+    # Litinski 2023 https://arxiv.org/abs/2306.08585
+    # Figure/Table 8. Lists n-qubit controlled modular addition as 5n toffoli.
+    #     Note: We have an extra toffoli due to how our OutOfPlaceAdder works.
+    assert counts['n_ccz'] == 5 * n + 1
+
+
 @pytest.mark.parametrize('control', range(2))
 @pytest.mark.parametrize(
     ['prime', 'bitsize'],
```

File: qualtran/bloqs/arithmetic/addition.py (hunk 1 above)
Comment (NoureldinYosri) [reply qualtran/bloqs/arithmetic/addition.py]: it's not about large $k$, it's about negative $k$. modular addition uses AddK with $-mod$ but the addition is unsigned so the call to QUInt fails validation. but adding -mod is the same as adding 2^bitsize - mod in this case ... I changed the code to explicitly apply this only for negative values and added a comment

File: qualtran/bloqs/mod_arithmetic/mod_addition_test.py (hunk 2 above)
Comment (NoureldinYosri) [reply qualtran/bloqs/mod_arithmetic/mod_addition_test.py:112]: done

File: qualtran/bloqs/mod_arithmetic/mod_addition_test.py (hunk 3 above)
Comment (NoureldinYosri) [reply qualtran/bloqs/mod_arithmetic/mod_addition_test.py]: done

File: qualtran/bloqs/mod_arithmetic/mod_addition.py (hunk 5 above)
Comment (NoureldinYosri) [reply qualtran/bloqs/mod_arithmetic/mod_addition.py]: updated

File: qualtran/bloqs/mod_arithmetic/mod_addition_test.py (hunk 6 above)
Comment (NoureldinYosri) [reply qualtran/bloqs/mod_arithmetic/mod_addition_test.py:103]: added

File: qualtran/bloqs/arithmetic/addition.py (hunk 1 above)
Comment (mpharrigan) [reply qualtran/bloqs/arithmetic/addition.py]: thanks! making these things "obvious" is very helpful

**Pull request merged**